File: genera-scheda.py
import os
import json
import zipfile
import base64
import tempfile
import anthropic
from http.server import BaseHTTPRequestHandler
import cgi

client = anthropic.Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))

SYSTEM_PROMPT = """Sei un esperto commercialista italiano specializzato in analisi di dichiarazioni dei redditi per underwriting.

Analizza la dichiarazione allegata ed estrai tutti i dati rilevanti. Poi genera una scheda cliente HTML completa usando ESATTAMENTE questo template HTML (sostituisci i segnaposto con i dati reali):

```html
<div class="scheda">
  <div class="scheda-header">
    <div>
      <div class="ragione-sociale">{{RAGIONE_SOCIALE}}</div>
      <div style="margin-top:0.5rem; display:flex; gap:0.5rem; flex-wrap:wrap;">
        <span class="badge badge-blue">{{TIPO_DICHIARAZIONE}}</span>
        <span class="badge badge-green">Anno {{ANNO_IMPOSTA}}</span>
      </div>
    </div>
    <div class="meta">
      <strong>{{CODICE_FISCALE}}</strong>
      P.IVA: {{PARTITA_IVA}}<br>
      Periodo: {{PERIODO_IMPOSTA}}<br>
      Presentata: {{DATA_PRESENTAZIONE}}
    </div>
  </div>

  <div class="section">
    <div class="section-title">Dati anagrafici</div>
    <div class="grid-2">
      <div class="field"><div class="field-label">Ragione sociale</div><div class="field-value">{{RAGIONE_SOCIALE}}</div></div>
      <div class="field"><div class="field-label">Codice fiscale</div><div class="field-value mono">{{CODICE_FISCALE}}</div></div>
      <div class="field"><div class="field-label">Forma giuridica</div><div class="field-value">{{FORMA_GIURIDICA}}</div></div>
      <div class="field"><div class="field-label">Rappresentante legale</div><div class="field-value">{{RAPPRESENTANTE}}</div></div>
      <div class="field"><div class="field-label">Telefono</div><div class="field-value">{{TELEFONO}}</div></div>
      <div class="field"><div class="field-label">Email</div><div class="field-value">{{EMAIL}}</div></div>
    </div>
  </div>

  {{SEZIONE_SOCI}}

  <div class="section">
    <div class="section-title">Dati reddituali</div>
    <div class="grid-3">
      <div class="field"><div class="field-label">Ricavi / Compensi</div><div class="amount">{{RICAVI}}</div></div>
      <div class="field"><div class="field-label">Altri proventi</div><div class="amount">{{ALTRI_PROVENTI}}</div></div>
      <div class="field"><div class="field-label">Totale componenti positivi</div><div class="amount">{{TOT_POSITIVI}}</div></div>
      <div class="field"><div class="field-label">Totale componenti negativi</div><div class="amount">{{TOT_NEGATIVI}}</div></div>
      <div class="field"><div class="field-label">Reddito d'impresa</div><div class="amount positive">{{REDDITO_IMPRESA}}</div></div>
      <div class="field"><div class="field-label">Codice attività ISA</div><div class="field-value mono">{{CODICE_ISA}}</div></div>
    </div>
  </div>

  <div class="section">
    <div class="section-title">Ripartizione tra soci</div>
    {{TABELLA_RIPARTIZIONE}}
  </div>

  <div class="section">
    <div class="section-title">Imposte e posizione fiscale</div>
    <div class="grid-2">
      <div class="highlight-row">
        <span class="label">Reddito imponibile netto</span>
        <span class="amount">{{REDDITO_IMPONIBILE}}</span>
      </div>
      <div class="highlight-row">
        <span class="label">Saldo a debito / credito</span>
        <span class="amount {{DEBITO_CREDITO_CLASS}}">{{DEBITO_CREDITO}}</span>
      </div>
    </div>
    <div class="grid-3" style="margin-top:0.75rem;">
      <div class="field"><div class="field-label">Quadri compilati</div><div class="field-value">{{QUADRI_COMPILATI}}</div></div>
      <div class="field"><div class="field-label">ISA allegato</div><div class="field-value">{{ISA}}</div></div>
      <div class="field"><div class="field-label">Incaricato trasmissione</div><div class="field-value">{{INCARICATO}}</div></div>
    </div>
  </div>

  <div class="section">
    <div class="section-title">Note e osservazioni</div>
    <div class="note-box">{{NOTE}}</div>
  </div>
</div>
```

Regole:
- Per {{SEZIONE_SOCI}}: se è una SNC/SAS/società di persone, includi una tabella soci con i dati del quadro RO. Usa la classe "soci-table".
- Per {{TABELLA_RIPARTIZIONE}}: tabella con colonne Socio | C.F. | Quota % | Reddito attribuito
- Importi: formatta sempre con € e separatore migliaia (es. € 32.399)
- Se un dato non è disponibile usa "n.d."
- Per {{DEBITO_CREDITO_CLASS}}: usa "positive" se credito, "negative" se debito
- Le NOTE devono essere sintetiche: tipo dichiarazione, situazioni particolari, eventuali perdite, superbonus, ecc.
- Rispondi SOLO con l'HTML puro, senza markdown, senza backtick, senza testo aggiuntivo.
"""


def extract_text_from_zip(zip_bytes):
    """Extract text files from zip containing jpeg+txt pages."""
    texts = []
    with tempfile.NamedTemporaryFile(suffix='.zip', delete=False) as f:
        f.write(zip_bytes)
        tmp_path = f.name

    with zipfile.ZipFile(tmp_path, 'r') as z:
        # Get all .txt files sorted by page number
        txt_files = sorted(
            [n for n in z.namelist() if n.endswith('.txt')],
            key=lambda x: int(x.replace('.txt', '')) if x.replace('.txt', '').isdigit() else 999
        )
        for name in txt_files:
            with z.open(name) as f:
                content = f.read().decode('utf-8', errors='replace')
                if content.strip():
                    texts.append(f"=== Pagina {name} ===\n{content}")

    os.unlink(tmp_path)
    return '\n'.join(texts)


def parse_multipart(environ):
    """Parse multipart form data."""
    content_type = environ.get('CONTENT_TYPE', '')
    content_length = int(environ.get('CONTENT_LENGTH', 0))
    body = environ['wsgi.input'].read(content_length)

    # Parse boundary
    boundary = None
    for part in content_type.split(';'):
        part = part.strip()
        if part.startswith('boundary='):
            boundary = part[9:].strip('"')
            break

    if not boundary:
        return None, None

    # Simple multipart parser
    boundary_bytes = ('--' + boundary).encode()
    parts = body.split(boundary_bytes)

    for part in parts[1:]:
        if b'name="file"' in part:
            # Extract filename
            filename = b''
            if b'filename="' in part:
                start = part.index(b'filename="') + 10
                end = part.index(b'"', start)
                filename = part[start:end]

            # Extract content (after double CRLF)
            if b'\r\n\r\n' in part:
                content = part.split(b'\r\n\r\n', 1)[1]
                # Remove trailing boundary marker
                if content.endswith(b'\r\n'):
                    content = content[:-2]
                return filename.decode('utf-8', errors='replace'), content

    return None, None


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            content_type = self.headers.get('Content-Type', '')
            body = self.rfile.read(content_length)

            # Parse multipart
            boundary = None
            for part in content_type.split(';'):
                part = part.strip()
                if part.startswith('boundary='):
                    boundary = part[9:].strip('"')
                    break

            filename = None
            file_content = None

            if boundary:
                boundary_bytes = ('--' + boundary).encode()
                parts = body.split(boundary_bytes)
                for part in parts[1:]:
                    if b'name="file"' in part:
                        if b'filename="' in part:
                            start = part.index(b'filename="') + 10
                            end = part.index(b'"', start)
                            filename = part[start:end].decode('utf-8', errors='replace')
                        if b'\r\n\r\n' in part:
                            content = part.split(b'\r\n\r\n', 1)[1]
                            if content.endswith(b'\r\n'):
                                content = content[:-2]
                            file_content = content
                        break

            if not file_content:
                self._json_error(400, "Nessun file ricevuto")
                return

            # Determine file type and extract text/images
            is_zip = file_content[:2] == b'PK'
            is_pdf = file_content[:4] == b'%PDF'

            messages_content = []

            if is_zip:
                # Extract text from zip pages
                extracted_text = extract_text_from_zip(file_content)
                messages_content.append({
                    "type": "text",
                    "text": f"Ecco il testo estratto dalle pagine della dichiarazione:\n\n{extracted_text}"
                })
            elif is_pdf:
                # Send PDF directly to Claude
                pdf_b64 = base64.standard_b64encode(file_content).decode('utf-8')
                messages_content.append({
                    "type": "document",
                    "source": {
                        "type": "base64",
                        "media_type": "application/pdf",
                        "data": pdf_b64
                    }
                })
            else:
                self._json_error(400, "Formato non supportato. Carica un PDF o un file ZIP.")
                return

            messages_content.append({
                "type": "text",
                "text": "Analizza questa dichiarazione e genera la scheda HTML come da istruzioni."
            })

            # Call Claude API
            response = client.messages.create(
                model="claude-sonnet-4-20250514",
                max_tokens=4096,
                system=SYSTEM_PROMPT,
                messages=[{"role": "user", "content": messages_content}]
            )

            html_content = response.content[0].text.strip()
            # Clean up any accidental markdown wrapping
            if html_content.startswith('```'):
                html_content = html_content.split('\n', 1)[1] if '\n' in html_content else html_content
                if html_content.endswith('```'):
                    html_content = html_content[:-3]
            html_content = html_content.strip()

            self._json_ok({"html": html_content})

        except anthropic.AuthenticationError:
            self._json_error(401, "API key Anthropic non valida o mancante")
        except Exception as e:
            self._json_error(500, f"Errore interno: {str(e)}")

    def _json_ok(self, data):
        body = json.dumps(data).encode()
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', len(body))
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(body)

    def _json_error(self, code, message):
        body = json.dumps({"error": message}).encode()
        self.send_response(code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', len(body))
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def log_message(self, format, *args):
        pass

File: api/genera-scheda.js
import Anthropic from "@anthropic-ai/sdk";
import { Redis } from "@upstash/redis";

export const config = { api: { bodyParser: false } };

const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
const redis = Redis.fromEnv();

const SYSTEM_PROMPT = `Sei un esperto commercialista italiano specializzato in analisi di dichiarazioni dei redditi per underwriting.

Analizza la dichiarazione allegata ed estrai tutti i dati rilevanti. Rispondi SOLO con un oggetto JSON (nessun testo aggiuntivo, nessun markdown) con questa struttura esatta:

{
  "ragione_sociale": "...",
  "codice_fiscale": "...",
  "partita_iva": "...",
  "tipo_dichiarazione": "...",
  "anno_imposta": "...",
  "periodo_imposta": "...",
  "data_presentazione": "...",
  "forma_giuridica": "...",
  "rappresentante": "...",
  "telefono": "...",
  "email": "...",
  "sede_legale": "...",
  "codice_isa": "...",
  "ricavi": "...",
  "altri_proventi": "...",
  "tot_positivi": "...",
  "tot_negativi": "...",
  "reddito_impresa": "...",
  "reddito_imponibile": "...",
  "debito_credito": "...",
  "debito_credito_segno": "credito",
  "quadri_compilati": "...",
  "isa_allegato": "Si",
  "incaricato": "...",
  "soci": [
    { "cf": "...", "nome": "...", "quota": "50%", "qualifica": "R", "reddito_attribuito": "..." }
  ],
  "note": "..."
}

Se un dato non e' disponibile usa "n.d.". Per debito_credito_segno usa "credito" o "debito". Gli importi vanno formattati con euro e separatore migliaia (es. "32.399").`;

async function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on("data", (c) => chunks.push(c));
    req.on("end", () => resolve(Buffer.concat(chunks)));
    req.on("error", reject);
  });
}

function parseBoundary(ct) {
  const m = ct.match(/boundary=([^\s;]+)/);
  return m ? m[1] : null;
}

function parseMultipart(body, boundary) {
  const delim = Buffer.from("--" + boundary);
  const parts = [];
  let start = 0;
  while (start < body.length) {
    const di = body.indexOf(delim, start);
    if (di === -1) break;
    const ps = di + delim.length;
    if (body[ps] === 45 && body[ps + 1] === 45) break;
    const hs = ps + 2;
    const he = body.indexOf(Buffer.from("\r\n\r\n"), hs);
    if (he === -1) break;
    const headers = body.slice(hs, he).toString();
    const cs = he + 4;
    const nd = body.indexOf(delim, cs);
    const ce = nd === -1 ? body.length : nd - 2;
    parts.push({ headers, content: body.slice(cs, ce) });
    start = nd === -1 ? body.length : nd;
  }
  return parts;
}

function buildHTML(d) {
  const sociSection = d.soci && d.soci.length > 0 ? `
    <div class="section">
      <div class="section-title">Soci</div>
      <table class="soci-table">
        <thead><tr><th>Codice fiscale</th><th>Nome</th><th>Quota</th><th>Qualifica</th></tr></thead>
        <tbody>${d.soci.map(s => `<tr><td class="cf">${s.cf}</td><td>${s.nome}</td><td>${s.quota}</td><td>${s.qualifica}</td></tr>`).join("")}</tbody>
      </table>
    </div>` : "";

  const ripartizione = d.soci && d.soci.length > 0 ? `
    <table class="soci-table">
      <thead><tr><th>Socio</th><th>C.F.</th><th>Quota %</th><th>Reddito attribuito</th></tr></thead>
      <tbody>${d.soci.map(s => `<tr><td>${s.nome}</td><td class="cf">${s.cf}</td><td>${s.quota}</td><td>${s.reddito_attribuito}</td></tr>`).join("")}</tbody>
    </table>` : "<p style='color:var(--text-muted);font-size:0.85rem;'>n.d.</p>";

  return `<div class="scheda">
  <div class="scheda-header">
    <div>
      <div class="ragione-sociale">${d.ragione_sociale}</div>
      <div style="margin-top:0.5rem;display:flex;gap:0.5rem;flex-wrap:wrap;">
        <span class="badge badge-blue">${d.tipo_dichiarazione}</span>
        <span class="badge badge-green">Anno ${d.anno_imposta}</span>
      </div>
    </div>
    <div class="meta">
      <strong>${d.codice_fiscale}</strong>
      P.IVA: ${d.partita_iva}<br>
      Periodo: ${d.periodo_imposta}<br>
      Presentata: ${d.data_presentazione}
    </div>
  </div>
  <div class="section">
    <div class="section-title">Dati anagrafici</div>
    <div class="grid-2">
      <div class="field"><div class="field-label">Ragione sociale</div><div class="field-value">${d.ragione_sociale}</div></div>
      <div class="field"><div class="field-label">Codice fiscale</div><div class="field-value mono">${d.codice_fiscale}</div></div>
      <div class="field"><div class="field-label">Forma giuridica</div><div class="field-value">${d.forma_giuridica}</div></div>
      <div class="field"><div class="field-label">Rappresentante legale</div><div class="field-value">${d.rappresentante}</div></div>
      <div class="field"><div class="field-label">Telefono</div><div class="field-value">${d.telefono}</div></div>
      <div class="field"><div class="field-label">Email</div><div class="field-value">${d.email}</div></div>
      <div class="field"><div class="field-label">Sede legale</div><div class="field-value">${d.sede_legale || "n.d."}</div></div>
    </div>
  </div>
  ${sociSection}
  <div class="section">
    <div class="section-title">Dati reddituali</div>
    <div class="grid-3">
      <div class="field"><div class="field-label">Ricavi / Compensi</div><div class="amount">${d.ricavi}</div></div>
      <div class="field"><div class="field-label">Altri proventi</div><div class="amount">${d.altri_proventi}</div></div>
      <div class="field"><div class="field-label">Tot. componenti positivi</div><div class="amount">${d.tot_positivi}</div></div>
      <div class="field"><div class="field-label">Tot. componenti negativi</div><div class="amount">${d.tot_negativi}</div></div>
      <div class="field"><div class="field-label">Reddito d'impresa</div><div class="amount positive">${d.reddito_impresa}</div></div>
      <div class="field"><div class="field-label">Codice attivita ISA</div><div class="field-value mono">${d.codice_isa}</div></div>
    </div>
  </div>
  <div class="section">
    <div class="section-title">Ripartizione tra soci</div>
    ${ripartizione}
  </div>
  <div class="section">
    <div class="section-title">Imposte e posizione fiscale</div>
    <div class="grid-2">
      <div class="highlight-row"><span class="label">Reddito imponibile netto</span><span class="amount">${d.reddito_imponibile}</span></div>
      <div class="highlight-row"><span class="label">Saldo a debito / credito</span><span class="amount ${d.debito_credito_segno === 'credito' ? 'positive' : 'negative'}">${d.debito_credito}</span></div>
    </div>
    <div class="grid-3" style="margin-top:0.75rem;">
      <div class="field"><div class="field-label">Quadri compilati</div><div class="field-value">${d.quadri_compilati}</div></div>
      <div class="field"><div class="field-label">ISA allegato</div><div class="field-value">${d.isa_allegato}</div></div>
      <div class="field"><div class="field-label">Incaricato trasmissione</div><div class="field-value">${d.incaricato}</div></div>
    </div>
  </div>
  <div class="section">
    <div class="section-title">Note e osservazioni</div>
    <div class="note-box">${d.note}</div>
  </div>
</div>`;
}

export default async function handler(req, res) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");
  if (req.method === "OPTIONS") return res.status(200).end();
  if (req.method !== "POST") return res.status(405).json({ error: "Method not allowed" });

  try {
    const body = await readBody(req);
    const contentType = req.headers["content-type"] || "";
    const boundary = parseBoundary(contentType);
    if (!boundary) return res.status(400).json({ error: "Nessun boundary trovato" });

    const parts = parseMultipart(body, boundary);
    let fileContent = null;
    for (const part of parts) {
      if (part.headers.includes('name="file"')) { fileContent = part.content; break; }
    }
    if (!fileContent || fileContent.length === 0) return res.status(400).json({ error: "Nessun file ricevuto" });

    const base64Content = fileContent.toString("base64");

    const response = await client.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 4096,
      system: SYSTEM_PROMPT,
      messages: [{
        role: "user",
        content: [
          { type: "document", source: { type: "base64", media_type: "application/pdf", data: base64Content } },
          { type: "text", text: "Analizza questa dichiarazione e rispondi SOLO con il JSON." }
        ]
      }]
    });

    let raw = response.content[0].text.trim();
    if (raw.startsWith("```")) {
      raw = raw.split("\n").slice(1).join("\n");
      if (raw.endsWith("```")) raw = raw.slice(0, -3);
    }

    const data = JSON.parse(raw.trim());
    const html = buildHTML(data);

    const id = `scheda:${Date.now()}`;
    await redis.set(id, JSON.stringify({
      id,
      ragione_sociale: data.ragione_sociale,
      codice_fiscale: data.codice_fiscale,
      anno_imposta: data.anno_imposta,
      tipo_dichiarazione: data.tipo_dichiarazione,
      html,
      created_at: new Date().toISOString()
    }));
    await redis.lpush("schede_index", id);

    return res.status(200).json({ html, id });
  } catch (error) {
    console.error("Error:", error);
    if (error.status === 401) return res.status(401).json({ error: "API key Anthropic non valida" });
    return res.status(500).json({ error: "Errore: " + error.message });
  }
}
